### PR TITLE
extend demo mode

### LINKED ISSUE #1078: extend demo mode
- make it available for everyone who is at least admin of 1 course
- display a banner when enabled (similar to impersonate)
- prevent saving user details when enabled

_[Original issue](https://github.ugent.be/dodona/dodona/issues/1078) by @bmesuere on Sat May 18 2019 at 11:11._

### PR DESCRIPTION
**Pull request by LucianosLionakis**
Fixes #1078

### CODE REVIEW

**Comment by LucianosLionakis** [comment]
Added banner when demo mode is active:
DeepinScreenshot_select-area_20190809133651

**Comment by LucianosLionakis** [comment]
Added icons to modes that show if mode is active or inactive:
DeepinScreenshot_select-area_20190809153150

File: app/assets/javascripts/index.js [hunk 1]
```diff
@@ -18,6 +18,19 @@ function setBaseUrl(_baseUrl) {
     window.dodona.index.doSearch();
 }
 
+function initModeButtons(demo_mode, darkmode){
```
Comment (bmesuere) [inline app/assets/javascripts/index.js]: Why do you need this code? Right now, both links trigger a page refresh so no need to update the icons dynamically.

If we would want to update checkboxes on other places, this code
- needs to move to another file
- drop the unused parameters in the function
- add a more specific class to target the links
- can be shortened by using toggleClass

File: app/views/layouts/_navbar.html.erb [hunk 2]
```diff
@@ -41,16 +41,32 @@
               <% end %>
               <% if policy(:pages).toggle_demo_mode? %>
                 <% if Current.demo_mode %>
-                  <li><%= link_to t(".disable_demo_mode"), toggle_demo_mode_path, remote: true, method: "POST" %></li>
+                  <li>
+                    <%= link_to toggle_demo_mode_path, remote: true, method: "POST" do %>
+                    <i class="mdi mdi-checkbox-marked-outline mdi-18"></i><%= t(".disable_demo_mode") %>
```
Comment (bmesuere) [inline app/views/layouts/_navbar.html.erb]: - Indent
- If we use a checkbox, the labels don't need to include "enable"/"disable" anymore. Simply use Dark mode and Demo mode

File: app/views/layouts/_navbar.html.erb [hunk 3]
```diff
@@ -82,3 +98,9 @@
     </div>
   </div>
 </nav>
+<script type="text/javascript">
```
Comment (bmesuere) [inline app/views/layouts/_navbar.html.erb]: You can also remove this if you delete the other function.

File: config/locales/views/defaults/en.yml [hunk 4]
```diff
@@ -13,6 +13,10 @@ en:
       impersonating_user: You are currently impersonating
       switch_back: Switch back
       your_account: to your account.
+    demo_mode:
```
Comment (bmesuere) [inline config/locales/views/defaults/en.yml]: Update text to something like **Demo mode enabled**: Dodona now conceals all real names. <Disable>.

File: config/locales/views/defaults/nl.yml [hunk 5]
```diff
@@ -13,6 +13,10 @@ nl:
       impersonating_user: Je speelt momenteel
       switch_back: Keer terug
       your_account: naar jouw account.
+    demo_mode:
```
Comment (bmesuere) [inline config/locales/views/defaults/nl.yml]: See English suggestion + s/mode/modus/

**Comment by LucianosLionakis** [comment]
Changed position of check-boxes so that they are at the left of the button:
DeepinScreenshot_select-area_20190810234735

File: config/locales/views/defaults/en.yml [hunk 6]
```diff
@@ -14,9 +16,8 @@ en:
       switch_back: Switch back
       your_account: to your account.
     demo_mode:
-      demo_mode_enabled: Your are currently in demo mode.
-      switch_back: Turn off
-      demo_mode_off: demo mode.
+      demo_mode_enabled: Demo mode enabled
```
Comment (bmesuere) [inline config/locales/views/defaults/en.yml]: Sorry if I wasn't clear (and a small additional tweak to the text):
**Demo mode enabled**: while in demo mode, Dodona conceals all users' names. [Disable demo mode](#).

File: config/locales/views/defaults/nl.yml [hunk 7]
```diff
@@ -14,9 +16,8 @@ nl:
       switch_back: Keer terug
       your_account: naar jouw account.
     demo_mode:
-      demo_mode_enabled: Je bent momenteel in demo mode.
-      switch_back: Schakel
-      demo_mode_off: demo mode uit.
+      demo_mode_enabled: Demo modus staat aan
```
Comment (bmesuere) [inline config/locales/views/defaults/nl.yml]: **Demo modus ingeschakeld**: in demo modus verbergt Dodona de echte namen van alle gebruikers. [Demo modus uitschakelen](#).

File: app/views/layouts/_navbar.html.erb [hunk 8]
```diff
@@ -41,16 +41,32 @@
               <% end %>
               <% if policy(:pages).toggle_demo_mode? %>
                 <% if Current.demo_mode %>
-                  <li><%= link_to t(".disable_demo_mode"), toggle_demo_mode_path, remote: true, method: "POST" %></li>
+                  <li>
+                    <%= link_to toggle_demo_mode_path, remote: true, method: "POST" do %>
+                    <i class="mdi mdi-checkbox-marked-outline mdi-18"></i><%= t(".demo_mode") %>
```
Comment (charvp) [inline app/views/layouts/_navbar.html.erb]: Indentation

File: app/views/layouts/_navbar.html.erb [hunk 9]
```diff
@@ -41,16 +41,32 @@
...
+                    <% end %>
+                  </li>
                 <% else %>
-                  <li><%= link_to t(".enable_demo_mode"), toggle_demo_mode_path, remote: true, method: "POST" %></li>
+                  <li>
+                    <%= link_to toggle_demo_mode_path, remote: true, method: "POST" do %>
+                    <i class="mdi mdi-checkbox-blank-outline mdi-18"></i> <%= t(".demo_mode") %>
```
Comment (charvp) [inline app/views/layouts/_navbar.html.erb]: Indentation

File: app/views/layouts/_navbar.html.erb [hunk 10]
```diff
@@ -41,16 +41,32 @@
...
+                    <% end %>
+                  </li>
                 <% end %>
               <% end %>
               <% if policy(:pages).toggle_dark_mode? %>
                 <% if session[:dark] %>
-                  <li><%= link_to t(".disable_dark_mode"), toggle_dark_mode_path, remote: true, method: "POST" %></li>
+                  <li>
+                    <%= link_to toggle_dark_mode_path, remote: true, method: "POST" do %>
+                    <i class="mdi mdi-checkbox-marked-outline mdi-18"></i> <%= t(".dark_mode") %>
```
Comment (charvp) [inline app/views/layouts/_navbar.html.erb]: Indentation

File: app/views/layouts/_navbar.html.erb [hunk 11]
```diff
@@ -41,16 +41,32 @@
...
+                    <% end %>
+                  </li>
                 <% else %>
-                  <li><%= link_to t(".enable_dark_mode"), toggle_dark_mode_path, remote: true, method: "POST" %></li>
+                  <li>
+                    <%= link_to toggle_dark_mode_path, remote: true, method: "POST" do %>
+                    <i class="mdi mdi-checkbox-blank-outline mdi-18"></i> <%= t(".dark_mode") %>
```
Comment (charvp) [inline app/views/layouts/_navbar.html.erb]: Indentation

File: public/javascripts/i18n.js [hunk 12]
```diff
@@ -107,15 +107,15 @@
     // Shift back
```
Comment (charvp) [inline public/javascripts/i18n.js]: This file shouldn't be in this PR.

**Pull request merged**